plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/InternalAwsS3Service.java

```diff
@@ -85,125 +86,106 @@ public synchronized AmazonS3 client(Settings repositorySettings) {
+                " must be accompanied by setting [" + S3Repository.Repository.KEY_SETTING + "]");
+        }
+        if (credentials == null) {
+            logger.debug("Using instance profile credentials");
+            return new PrivilegedInstanceProfileCredentialsProvider();
+        } else {
+            logger.debug("Using basic key/secret credentials");
+            return new StaticCredentialsProvider(credentials);
         }
     }
 
     // pkg private for tests
     /** Returns the endpoint the client should use, based on the available endpoint settings found. */
-    static String findEndpoint(Logger logger, Settings repositorySettings, Settings settings, String clientName) {
-        String endpoint = getConfigValue(repositorySettings, settings, clientName, S3Repository.ENDPOINT_SETTING,
-                                         S3Repository.Repository.ENDPOINT_SETTING, S3Repository.Repositories.ENDPOINT_SETTING);
-        if (Strings.isNullOrEmpty(endpoint)) {
-            // No region has been set so we will use the default endpoint
-            if (CLOUD_S3.ENDPOINT_SETTING.exists(settings)) {
-                endpoint = CLOUD_S3.ENDPOINT_SETTING.get(settings);
-                logger.debug("using explicit s3 endpoint [{}]", endpoint);
-            }
-        } else {
+    static String findEndpoint(Logger logger, S3ClientSettings clientSettings, Settings repositorySettings) {
+        String endpoint = getRepoValue(repositorySettings, S3Repository.Repository.ENDPOINT_SETTING, clientSettings.endpoint);
+        if (Strings.isNullOrEmpty(endpoint) == false) {
             logger.debug("using repository level endpoint [{}]", endpoint);
         }
-
         return endpoint;
     }
 
-    /**
-     * Find the setting value, trying first with named configs,
-     * then falling back to repository and global repositories settings.
-     */
-    private static <T> T getConfigValue(Settings repositorySettings, Settings globalSettings, String clientName,
-                                        Setting.AffixSetting<T> configSetting, Setting<T> repositorySetting, Setting<T> globalSetting) {
-        Setting<T> concreteSetting = configSetting.getConcreteSettingForNamespace(clientName);
-        if (concreteSetting.exists(globalSettings)) {
-            return concreteSetting.get(globalSettings);
-        } else if (repositorySetting == null) {
-            // no repository setting, just use global setting
-            return globalSetting.get(globalSettings);
-        } else {
-            return getValue(repositorySettings, globalSettings, repositorySetting, globalSetting);
+    /** Returns the value for a given setting from the repository, or returns the fallback value. */
+    private static <T> T getRepoValue(Settings repositorySettings, Setting<T> repositorySetting, T fallback) {
+        if (repositorySetting.exists(repositorySettings)) {
```

Comment:
It's too bad our settings don't mimic the `Map` `getOrDefault` to make this one operation